distribution/push_v2.go

```diff
@@ -263,6 +265,37 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
 	// then push the blob.
 	bs := pd.repo.Blobs(ctx)
 
+	// attempt to find another repository to mount the layer from
+	for _, blobsum := range possibleBlobsums {
+		sourceRepo, err := reference.ParseNamed(blobsum.SourceRepository)
+		if err != nil {
+			continue
+		}
+		if pd.repoInfo.Hostname() == sourceRepo.Hostname() {
+			logrus.Debugf("attempting to mount layer %s (%s) from %s", diffID, blobsum.Digest, sourceRepo.FullName())
+
+			desc, err := bs.Mount(ctx, sourceRepo.RemoteName(), blobsum.Digest)
+			if err == nil {
+				progress.Updatef(progressOutput, pd.ID(), "Mounted from %s", sourceRepo.RemoteName())
```

Comment:
An error is currently returned by the client on `StatusAccepted`, but we cache the blob upload information for the next call to `Create`